### PR TITLE
Function Add LipSyncAnimaion And BlinkAnimation

### PR DESCRIPTION
**Pull request by generalfreed**
Click here for details
https://github.com/MakinoharaShoko/WebGAL/issues/344

### CODE REVIEW

**Review by MakinoharaShoko — CHANGES_REQUESTED** [review]
Using a large number of fields in the state table, if it is for a separate function, why not refer to the "effect" field for processing? The "effect" field expresses how to add effects to a specific character or background image at a certain ID or position. Can't we use the same approach for the mouth movement and blinking effects here?

**Review by MakinoharaShoko — CHANGES_REQUESTED** [review]
It is not necessary to write a function called getFigureByKey. Instead, consider adding a parameter to the existing getObjectByKey function. This way, if you only need to find an object from the figure, you can do so

**Review by MakinoharaShoko — CHANGES_REQUESTED** [review]
Since we have already added the "Associated ID" feature in the previous version, what is the purpose of the dispatch figurePos here? For illustrations that only set the position without setting an ID, we use fig-left, fig-center, and fig-right to represent the ID

**Review by MakinoharaShoko — CHANGES_REQUESTED** [review]
The state could be considered to change to this form:
```ts
interface IFigureAssociatedAnimation{
  mouthAnimation: IMouthAnimationFile;
  blinkAnimation: IEyesAnimationFile;
  targetId: string;
  flag:string;
}

// stageState
interface IStageState{
  ......
  figureAssociatedAnimation: IFigureAssociatedAnimation;
  ......
}
```

**Review by MakinoharaShoko — CHANGES_REQUESTED** [review]
If there is no associated mouth movement or blinking animation in the figure, is dispatch still necessary? Do we need to make a conditional judgment?

**Comment by generalfreed** [comment]
@MakinoharaShoko 
Thank you for the code review and valuable suggestions. I will try to address your concerns as follows:

**Regarding the use of the effect field:**

Are you referring to what's located in \webgal\src\Components\Stage\MainStage\useSetEffects.ts? If so, I will look into leveraging this existing functionality for the mouth movement and blinking effects.

**On replacing getFigureByKey with getObjectByKey:**

I tried to search for the term "getObjectByKey" in the codebase but couldn't find it. Could you please specify where it's being used so that I can make the necessary adjustments?

**On the purpose of dispatch figurePos and the "Associated ID":**

I've also considered the case where only the position is specified without an ID. If it's deemed unnecessary, I can switch to relying solely on IDs.

**Conditionally executing dispatch:**

I plan on adding a conditional check before executing the dispatch to make sure it's only done when necessary.

Please let me know if these changes are in line with your expectations or if there's anything else that needs to be addressed.

**Comment by MakinoharaShoko** [comment]
1. It does not rely entirely on ID. I don't know if you've noticed, but a certain position's illustration will be assigned a default ID, such as "fig-left" for the left side.
2. There is a "getStageObjByKey" function in PixiController
3. It's not about using the "effect" field, but rather referring to it. Use an array (where each item describes an object that animates the mouth, eyes, etc. for a specific ID) to describe the effects of the illustration

File: packages/webgal/src/store/stageInterface.ts
Comment (MakinoharaShoko) [inline packages/webgal/src/store/stageInterface.ts]: Is this necessary here?

File: packages/webgal/src/Core/gameScripts/changeFigure.ts
Comment (MakinoharaShoko) [inline packages/webgal/src/Core/gameScripts/changeFigure.ts]: When the illustration is switched or unloaded, the mouth shape or blink animation associated with that key should be deleted.
In addition, the handling here is strange. Shouldn't it be to first retrieve the previous array, make some modifications, and then dispatch? Creating a new array here implies that if another illustration was previously set with mouth shape and blink animation, it will be cleared

File: packages/webgal/src/Core/gameScripts/changeFigure.ts
Comment (MakinoharaShoko) [inline packages/webgal/src/Core/gameScripts/changeFigure.ts]: Consider switching to "switch", there are too many "if" statements

File: packages/webgal/src/store/stageInterface.ts
Comment (generalfreed) [reply packages/webgal/src/store/stageInterface.ts]: Delete the following as it is unnecessary
`animationFlag: string;`

Comment (MakinoharaShoko) [reply packages/webgal/src/store/stageInterface.ts]: What I mean is,
figureId: string;
figurePos: string;
It may not be necessary, as there is a key in IFigureAssociatedAnimation

Comment (MakinoharaShoko) [reply packages/webgal/src/store/stageInterface.ts]: Moreover, figureId and figurePos are not clearly defined

**Comment by MakinoharaShoko** [comment]
You're overcomplicating things; take it easy. I think to implement your feature, you only need to do the following:

No need to modify playVocal, because when the vocal is playing, we can use useEffect to monitor the current vocal in the state table. If it changes, that means the vocal is being played, right? In earlier versions, we've already associated the vocal with the character target.

All you need to do when setting up the figure (character illustration) is allow for images related to lip movements and blinking. Just use IFigureAssociatedAnimation to record it, then update the array in the state. Why an array? Because we can have more than one character on the field.

The specific approach is to check the figureAssociatedAnimation: Array<IFigureAssociatedAnimation> array when you detect a change in currentVocal. If the current vocal-associated figure isn't in this list, do nothing. If it's in the list, execute the corresponding animation (like for the mouth movements).

When changing or clearing a figure (setting it to 'none'), actually, every time a figure at a certain position or ID changes, first remove that figure's mouth animation effect from the state table. What if we need to add effects at this time? That's not an issue. When the figure leaves or switches, to prevent any mismatch between the figure and effects, we need to clear the effects. If you don't understand what I'm saying, take a look at how I handled it [here](https://github.com/MakinoharaShoko/WebGAL/blob/a593b39d5acd9c9a21f6541bbb1e4c6fdc466aff/packages/webgal/src/Core/gameScripts/changeFigure.ts#L47-L57).

**Comment by generalfreed** [comment]
> You're overcomplicating things; take it easy. I think to implement your feature, you only need to do the following:
> 
> No need to modify playVocal, because when the vocal is playing, we can use useEffect to monitor the current vocal in the state table. If it changes, that means the vocal is being played, right? In earlier versions, we've already associated the vocal with the character target.
> 
> All you need to do when setting up the figure (character illustration) is allow for images related to lip movements and blinking. Just use IFigureAssociatedAnimation to record it, then update the array in the state. Why an array? Because we can have more than one character on the field.
> 
> The specific approach is to check the figureAssociatedAnimation: Array array when you detect a change in currentVocal. If the current vocal-associated figure isn't in this list, do nothing. If it's in the list, execute the corresponding animation (like for the mouth movements).
> 
> When changing or clearing a figure (setting it to 'none'), actually, every time a figure at a certain position or ID changes, first remove that figure's mouth animation effect from the state table. What if we need to add effects at this time? That's not an issue. When the figure leaves or switches, to prevent any mismatch between the figure and effects, we need to clear the effects. If you don't understand what I'm saying, take a look at how I handled it [here](https://github.com/MakinoharaShoko/WebGAL/blob/a593b39d5acd9c9a21f6541bbb1e4c6fdc466aff/packages/webgal/src/Core/gameScripts/changeFigure.ts#L47-L57).

I don't understand what you mean, so me put in order.

you opinion is that the following variables are unnecessary?
```
figureId: string;
figurePos :string;
```

These are so that you can individually specify the figures (character illustrations)  want to animate blinking and lip-syncing.
Conversely, the targetId of IFigureAssociatedAnimation added to discrimination the figure (character illustration) to be animated.

In other words, when the figure you want to blink or lip-sync (character illustration) matches the figure want to animate (character illustration)
Animation is performed for the first time.

Is this all right?

Next is the method of clearing blinking and lip-syncing animations, but it seems impossible to do everything with changeFigure, so
Referring to the Effects processing, add processing like removeAnimationWithSetAnimations in useSetFigure.ts, then does it seem better to do the processing with changeFigure?

おっしゃてる意味がわからないので、整理させてください。

Mahiruさんの意見は、下記の変数は不要とのことですか？
figureId: string;
figurePos :string;

これらは、まばたき、口パクのアニメーションしたいフィギュア（キャラクターイラスト）を個別に指定できるようにするため
追加しました。
逆に、IFigureAssociatedAnimationのtargetIdは、アニメーション対象となるフィギュア（キャラクターイラスト）を判別するために追加しました。

つまり、まばたき、口パクしたいフィギュア（キャラクターイラスト）と、アニメーション対象となるフィギュア（キャラクターイラスト）が一致したときに
初めてアニメーションが行われます。

ここまでの前提に認識齟齬はないですか？

次に、まばたき、口パクアニメーションクリアを行う方法ですが、changeFigureですべてを行うのは、無理がありそうなので、
Effectsの処理を参考に、useSetFigure.tsでremoveAnimationWithSetAnimationsのような処理を追加して、それからchangeFigureで処理を行うのがよさそうですか？

**Comment by MakinoharaShoko** [comment]
```
figureId: string;
figurePos: string;
```

We don't need these two variables to infer which figure requires lip sync. The method is as follows:
1. Lip sync is always triggered by playVocal, and playVocal is always triggered by say. However, in say, we can specify the associated ID or the character at a specific position
2. Therefore, at the beginning of the voice, we can infer from above which key or position the character illustration should be synchronized with.
3. In the variable figureAssociatedAnimation: Array<IFigureAssociatedAnimation>, we can use the find method to attempt to find out if there is a lip sync configuration that matches this key or position for the character illustration. If there is, synchronize it; if not, do nothing

**Comment by MakinoharaShoko** [comment]
When I say "clearing lip sync configuration," I mean that when using changeFigure, you should first remove the **specific item associated with the upcoming figure to be set** from the `figureAssociatedAnimation: Array<IFigureAssociatedAnimation>` array, without altering the other items in the array

**Comment by MakinoharaShoko** [comment]
https:github.com/MakinoharaShoko/WebGAL_Terre/blob/535dd6a7407895b88db3d61d5df0c277d5b7d812/packages/origine2/src/pages/editor/GraphicalEditor/SentenceEditor/Say.tsx#L59C1-L59C1

You added an associated figureId or figurePosition to the Say command here. In this feature, you can use these two parameters to specify the target for lip synchronization with this sentence.

**Comment by MakinoharaShoko** [comment]
Translation to English: I don't know if you've noticed, but whether you specify a location or a key, the end result is always specifying a key. Specifying a location just forms a special key. I looked at your implementation, and the blinking animation doesn't use the keys in the state table at all. As for lip-syncing, since you need to get the voice, you chose to handle it in the AudioContainer. But do we really need to handle it there? Starting at line 90 in playVocal.ts, we have the logic for playing the audio after the voice files are prepared, as well as how to handle it when the audio ends. I think it's more appropriate to handle the lip-syncing logic here, starting the lip animation when 'oncanplay' is triggered and stopping it when 'onended' is triggered. Wouldn't this be more foolproof and align better with the main performance logic? You can also know which character illustration needs lip-syncing from the sentence parameter in playVocal.

**Comment by generalfreed** [comment]
@MakinoharaShoko 

Clear Lip-Sync Settings: Implemented logic to clear the lip-sync settings for the next figure that will be displayed. This involved selectively removing associated animation items from the figureAssociatedAnimation array without affecting other items in the array.

Moved Functionality: Transferred audio level and lip-sync related functionalities from AudioContainer.tsx to playVocal.ts.

Code Cleanup: Removed unused variables 
```
figureId: string;
figurePos: string;
```

**Comment by generalfreed** [comment]
@MakinoharaShoko 
>Lack of necessary debugging materials, please check the files under public/game
Should I update the demo_animation.txt file for lip sync and blink animation as part of the debugging materials?
Could you clarify what files under public/game need to be updated?
Is demo_animation.txt related to this?

**Comment by MakinoharaShoko** [comment]
> 



> @MakinoharaShoko
> 
> > Lack of necessary debugging materials, please check the files under public/game
> 
> Should I update the demo_animation.txt file for lip sync and blink animation as part of the debugging materials? Could you clarify what files under public/game need to be updated? Is demo_animation.txt related to this?

Image assets.

**Comment by generalfreed** [comment]
@MakinoharaShoko 
Added.
Please check demo_animation.txt.

**Review by MakinoharaShoko — APPROVED** [review]

**Comment by MakinoharaShoko** [comment]
I have made some refactoring to the code structure and addressed a problem with lip synchronization. However, I am unsure if the blinking animation is functioning properly after the refactoring. Please verify

**Comment by generalfreed** [comment]
@MakinoharaShoko 
Thank you for your thorough review and for the refactoring you've undertaken, particularly for fixing the lip synchronization issue. I have also verified the blinking animation functionality through testing, and it appears to be working as expected. 

Is there anything else you think we should address before merging this PR?

**Comment by MakinoharaShoko** [comment]
Not at the moment, we will merge after conducting some further checks in the near future

**Pull request merged**
